File: src/019RemoveNthNodeFromEndOfList/019RemoveNthNodeFromEndOfList.go
package main

/**
 * Definition for singly-linked list.
 **/
type ListNode struct {
	Val  int
	Next *ListNode
}

func removeNthFromEnd(head *ListNode, n int) *ListNode {
	if n <= 0 {
		return head
	}
	dummy := ListNode{0, head}
	p := &dummy
	q := &dummy
	for i := 0; i < n; i++ {
		q = q.Next
	}
	for q.Next != nil {
		q = q.Next
		p = p.Next
	}
	p.Next = p.Next.Next
	return dummy.Next
}

func main() {

}

File: src/019RemoveNthNodeFromEndOfList/019RemoveNthNodeFromEndOfList.rs
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if (n <= 0) {
            return head;
        }
        
        ListNode dummy = ListNode(0);
        dummy.next = head;
        ListNode *r = &dummy;
        ListNode *p = dummy.next;
        ListNode *q = dummy.next;
                
        while (n-- > 0 && q != nullptr) {
            q = q->next;
        }
        
        while (q!= nullptr) {
            r = r->next;
            p = p->next;
            q = q->next;
        }
        
        r->next = p->next;        
        
        return dummy.next;
    }
};
